From b619fc942c1f262395da2d87691cb0d871841da7 Mon Sep 17 00:00:00 2001
From: Varun Sharma <vsharma@habana.ai>
Date: Thu, 18 Mar 2021 14:46:32 +0200
Subject: [PATCH] [SW-24964] i2c indirect: add support for smbus alert opcode(8)

This patch support read of smbus alert opcode 8 via
indirect protocol from bmc. This will be used to report fatal
errors like hbm init failure, sram ecc double error during previous boot.

Change-Id: I82d4bc321b6be470d03a838b6d3a461a716636ea
Signed-off-by: Varun Sharma <vsharma@habana.ai>
---

diff --git a/gaudi/rma_info.c b/gaudi/rma_info.c
index 59c580e..c02367a 100644
--- a/gaudi/rma_info.c
+++ b/gaudi/rma_info.c
@@ -14,6 +14,7 @@
 #include "boot_if.h"
 #include "module_i2c_slave.h"
 #include "hl_i2c_slave_events.h"
+#include "i2cslave_indirect.h"
 
 #define ROUND_UP4(n)	(((n) + sizeof(u32) - 1) & ~(sizeof(u32) - 1))
 #define RMA_HEADER_ENTRY_SIZE		sizeof(struct rma_entry)
@@ -175,6 +176,7 @@
 		rma_header->rma_failure.row_replacement ||
 		rma_header->rma_failure.hbm_init) {
 		err_sts_reg_val |= HL_ERROR_STATUS_FATAL_ERROR;
+		i2csi_update_smbus_alert(rma_header);
 		/* set boot error bit for RMA */
 		cpu_boot_err_clrset(CPU_BOOT_ERR0_DEVICE_UNUSABLE_FAIL, true);
 		preboot_puts_hex("Fatal error:", rma_header->rma_failure.data, 0);
diff --git a/include/i2cslave_indirect.h b/include/i2cslave_indirect.h
index 2ad5023..c63660f 100644
--- a/include/i2cslave_indirect.h
+++ b/include/i2cslave_indirect.h
@@ -9,6 +9,7 @@
 #define _I2CSLAVE_INDIRECT_H_
 
 #include <hl_fw_if.h>
+#include <rma_if.h>
 
 u8 i2csi_req_handle_sts(u8 cmd);
 u8 i2csi_req_handle_rdata(u8 cmd, u32 read_seq);
@@ -24,5 +25,5 @@
 bool i2csi_ctrl_is_rst_dev(void);
 void i2csi_set_boot_err(enum hl_fw_fit_boot_error err,
 			bool is_linux_boot_fail);
-
+void i2csi_update_smbus_alert(struct rma_entry *rma_header);
 #endif /* _I2CSLAVE_INDIRECT_H_ */
diff --git a/preboot/i2cslave_indirect.c b/preboot/i2cslave_indirect.c
index fa4a7e0..cd7e015 100644
--- a/preboot/i2cslave_indirect.c
+++ b/preboot/i2cslave_indirect.c
@@ -16,11 +16,13 @@
 #include <crc32.h>
 #include <hl_fw_if.h>
 #include "fit.h"
+#include "i2cslave_indirect.h"
 
 #define SUPPORTED_OPCODES_MASK	( \
 				BIT(HL_I2CI_REQ_OP_DEVICE_INFO) | \
 				BIT(HL_I2CI_REQ_OP_DEVICE_STATUS) | \
 				BIT(HL_I2CI_REQ_OP_CONFIG_ACCESS) | \
+				BIT(HL_I2CI_REQ_OP_SMBUS_ALERT) | \
 				BIT(HL_I2CI_REQ_OP_FW_LOADER) | \
 				BIT(HL_I2CI_REQ_OP_MEMCPY_CTRL) | \
 				BIT(HL_I2CI_REQ_OP_MEMCPY_DATA) | \
@@ -100,6 +102,7 @@
 static struct hl_i2ci_memcpy_ctrl op_memcpy __aligned(8);
 
 static struct hl_i2ci_error_status op_err_status;
+static struct hl_i2ci_smbus_alert smbus_alert;
 
 static struct hl_i2ci_ctrl_cmd ctrl_cmd;
 static struct hl_i2ci_ctrl_sts ctrl_sts;
@@ -338,6 +341,10 @@
 	case HL_I2CI_REQ_OP_CONFIG_ACCESS:
 		req_rd.data = (u8 *)&op_config;
 		req_rd.length = sizeof(op_config);
+		break;
+	case HL_I2CI_REQ_OP_SMBUS_ALERT:
+		req_rd.data = (u8 *)&smbus_alert;
+		req_rd.length = sizeof(smbus_alert);
 		break;
 	case HL_I2CI_REQ_OP_MEMCPY_CTRL:
 		req_rd.data = (u8 *)&op_memcpy;
@@ -794,6 +801,22 @@
 		ctrl_cmd.opcode == HL_I2CI_CTRL_OP_RESET_DEVICE;
 }
 
+void i2csi_update_smbus_alert(struct rma_entry *rma_header)
+{
+	if (rma_header->rma_failure.sram_ecc)
+		smbus_alert.fatal_error |= HL_I2CI_SMBUS_FATAL_ERROR_SRAM_ECC;
+	if (rma_header->rma_failure.hbm_ecc)
+		smbus_alert.fatal_error |= HL_I2CI_SMBUS_FATAL_ERROR_HBM_ECC;
+	if (rma_header->rma_failure.hbm_parity)
+		smbus_alert.fatal_error |=
+				HL_I2CI_SMBUS_FATAL_ERROR_HBM_PARITY;
+	if (rma_header->rma_failure.row_replacement)
+		smbus_alert.fatal_error |=
+				HL_I2CI_SMBUS_FATAL_ERROR_ROW_REPLACEMENT;
+	if (rma_header->rma_failure.hbm_init)
+		smbus_alert.fatal_error |= HL_I2CI_SMBUS_FATAL_ERROR_HBM_INIT;
+}
+
 void i2csi_set_boot_err(enum hl_fw_fit_boot_error err, bool is_linux_boot_fail)
 {
 	struct hl_i2ci_error_status_fit_failure *fit_fail;
